CRFModuleValidation/MasterScheduledActivityManager.swift

```diff
@@ -0,0 +1,563 @@
+//        if let first = futureSchedules.first,
+//            let firstIndex = newSections.index(where: { $0.date == first.date }) {
+//
+//            var replaceSchedules = futureSchedules
+//            var fromIndex = firstIndex
+//
+//            // Look at today and add the new schedule if it isn't already added for today
+//            let todayWeekday = Calendar.gregorian.component(.weekday, from: self.today)
+//            if newSchedule.daysOfWeek.contains(todayWeekday),
+//                let todaySection = newSections.first(where: { $0.date.isToday }),
+//                !todaySection.contains(taskGroup: newSchedule.taskGroup) {
+//                let newItem = ScheduleItem(date: self.today, taskGroup: newSchedule.taskGroup, isCompleted: false)
+//                let newTodaySection = ScheduleSection(items: todaySection.items.appending(newItem))
+//                replaceSchedules.insert(newTodaySection, at: 0)
+//                fromIndex = fromIndex - 1
+//            }
+//
+//            // replace future and today sections as appropriate
+//            newSections.replaceSubrange(fromIndex..<newSections.endIndex, with: replaceSchedules)
+//        }
+//
+//        // Update the schedules and sections and reset the notifications
+//        updateSchedules(newSchedules: newSchedules, newSections: newSections, shouldResetNotifications: true)
+//    }
+    
+    func updateReminderNotifications() {
+        
+        // use dispatch async to allow the method to return and put updating reminders on the next run loop
+        DispatchQueue.main.async {
+            
+            // Remove previous reminders
+            let identifiers = self.schedules.map({ $0.taskGroup.identifier })
+            UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: identifiers)
+            
+            // Check that there are any reminders to set and otherwise, do not even check for permission
+            guard self.schedules.filter({ $0.timeOfDay != nil }).count > 0 else { return }
+
+            // Check for permission and if granted, then schedule the reminders
+            if let settings = UIApplication.shared.currentUserNotificationSettings, settings.types != [] {
+                self.addLocalNotifications()
+            }
+            else {
+                UNUserNotificationCenter.current().requestAuthorization(options: [.badge, .alert, .sound]) { (granted, _) in
+                    if granted {
+                        self.addLocalNotifications()
+                    }
+                }
+            }
+        }
+    }
+    
+    fileprivate func addLocalNotifications() {
+        DispatchQueue.main.async {
+            for schedule in self.schedules {
+                schedule.scheduleReminder()
+            }
+        }
+    }
+
+    func createTaskViewController(for taskIdentifier: TaskIdentifier) -> RSDTaskViewController? {
```

Comment:
Look at the didSelect code. This would break for Bridge Surveys which are still using ResearchKit.

Reply:
This is used for presenting task view controllers directly from the Journey timeline, and those are only either 12 minute run or Stair step. Surveys in CRF are only ever presented from a TaskGroupTableViewController during one of the clinic visits, which uses a GroupedScheduledActivityManager whose didSelect function now falls through to ScheduledActivityManager.